### PR TITLE
Validate timeout and timeunit values

### PR DESCRIPTION
**Pull request by rozza**
Ensure that users can't accidently set a 0 value
which in CSOT represents infinite timeouts.

JAVA-5351

### CODE REVIEW

File: driver-core/src/test/unit/com/mongodb/ClientSessionOptionsSpecification.groovy [hunk 1]
```diff
@@ -18,6 +18,8 @@ package com.mongodb
 
 import spock.lang.Specification
 
+import java.util.concurrent.TimeUnit
+
 class ClientSessionOptionsSpecification extends Specification {
```
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/ClientSessionOptionsSpecification.groovy:23]: This class was also moved from the drivers-legacy dir. There seemed to be no reason to have it there as it tests driver-core functionality.

File: driver-core/src/test/unit/com/mongodb/ClientEncryptionSettingsSpecification.groovy [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+package com.mongodb
+
+import spock.lang.Specification
+
+import javax.net.ssl.SSLContext
+import java.util.concurrent.TimeUnit
+
+class ClientEncryptionSettingsSpecification extends Specification {
```
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/ClientEncryptionSettingsSpecification.groovy:24]: There was no unit tests for this class - so added this for extra coverage.

File: driver-core/src/main/com/mongodb/ClientSessionOptions.java [hunk 3]
```diff
@@ -224,8 +224,9 @@ public Builder defaultTransactionOptions(final TransactionOptions defaultTransac
          * @see #getDefaultTimeout
          */
         public Builder defaultTimeout(final long defaultTimeout, final TimeUnit timeUnit) {
-            isTrueArgument("timeoutMS must be >= 0", defaultTimeout >= 0);
-            this.defaultTimeoutMS = MILLISECONDS.convert(defaultTimeout, timeUnit);
+            this.defaultTimeoutMS = isTrueArgument("timeoutMS must be >= 0",
```
Comment (vbabanin) [inline driver-core/src/main/com/mongodb/ClientSessionOptions.java:227]: To enhance clarity, we could update the validation message to be more descriptive, specifying the parameter name and differentiating between the two scenarios where validation fails. Currently, we use a single message for both cases.

I propose two distinct messages:
For a too small defaultTimeout: `defaultTimeout is too small. It was rounded to 0 milliseconds, resulting in an unintended infinite timeout.`
For a negative value after conversion: `defaultTimeout, after conversion to milliseconds, must be >= 0.`

This suggestion applies to all timeout/withTimeout methods.
What do you think?

**Review by vbabanin — COMMENTED** [review]
Looks good!

One thing: I think we should add the same validation to the `withTimeout` methods of `MongoDatabase`, `MongoCollection`, and `GridFSBucket`.

File: driver-core/src/main/com/mongodb/ClientSessionOptions.java (hunk 3 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/ClientSessionOptions.java]: SGTM!

**Comment by rozza** [comment]
Doh! My search missed all the withTimeout methods - thanks.

**Review by vbabanin — APPROVED** [review]

**Pull request merged**